### DISCUSSION TITLE
Unofficial specs for Github Copilot plugin-agent.js interactions

### DISCUSSION THREAD

**Discussion by haukot** [zbirenbaum/copilot.lua#184] [General]
HI! I've done research and reverse-engineered the Github Copilot protocol between IDE plugin and agent.js, and have tried to document it. This is based on Github Copilot plugin version for JetBrains and some parts of agent.js. You can see the specs in this repo https://github.com/haukot/copilot_specifications
Hope this could help in plugin development!

**Comment by MunifTanjim** [comment]
Nice work!

We have these reverse-engineered APIs in this file:
https://github.com/zbirenbaum/copilot.lua/blob/master/lua/copilot/api.lua

They're mostly taken from the official plugin:
https://github.com/github/copilot.vim

Would be really nice if somebody reverse-engineers the new Copilot Chat features: https://marketplace.visualstudio.com/items?itemName=GitHub.copilot-chat. Currently it's only available on VSCode.

**Comment by haukot** [comment]
@MunifTanjim
Yes, that's great! There are some differences between Vim and JetBrains versions that, in my understanding, are valuable: commands "textDocument/didFocus"(to have better context based on last access time), "textDocument/didChange" with range(so, it'll send only the changed part of the document to agent.js on each request.), and "textDocument/didClose"(to remove unused files from current context).

>Would be really nice if somebody reverse-engineers the new Copilot Chat features: https://marketplace.visualstudio.com/items?itemName=GitHub.copilot-chat. Currently it's only available on VSCode.

It is harder because it's necessary either to make fixes directly in the extension.js code (which is probably legally restricted by the license), or to write all the code by hand (which is likely a lot of work).
I think a promising approach might be to create a wrapper that will implement methods from VS Code that the plugin is using, run extension.js, and provide an API similar to that of agent.js.
VS Code itself is open-source, and perhaps not many functions are used in Copilot Chat, so it could be possible to implement.

In this way, we don't need to reverse engineer all the code; we only need to understand which functions and callbacks of VS Code the plugin is using.

**Reply by MunifTanjim** [reply]
> "textDocument/didChange" with range(so, it'll send only the changed part of the document to agent.js on each request.), and "textDocument/didClose"(to remove unused files from current context).

Those 2 commands are sent by Neovim itself. Somewhere in this file: https://github.com/neovim/neovim/blob/master/runtime/lua/vim/lsp.lua

**Reply by zbirenbaum** [reply]
Using JetBrains is the way to go. The older versions are even more useful because they didn't have an agent.js, all the auth logic was handled in the jetbrains plugin itself instead of the almost illegible agent.js. Java decompilers give a much more legible output than the js deobfuscators. The biggest challenge at the moment is the inability to authenticate from a 3rd party client. I've tried my hand at this a number of times but it's quite difficult.

The only  reasonable way to support the experimental features + add improvements etc would be to write our own agent.js equivalent as you stated.

**Reply by haukot** [reply]
@zbirenbaum 
>The older versions are even more useful because they didn't have an agent.js

Interesting, didn't know that!

I actually thought about own agent.js from a different point of view - as a wrapper above extension.js from VS Code. Because there is a quite of work with prompt preparation and detection of the best completion, and it is happening in agent.js(or extension.js).
And it could be a lot of work to be in sync with Copilot internal API too.

But you are right that if we'll call Copilot API directly we can get much more flexibility.

So, if it'd help with auth, I managed to get it, you can check the example here https://github.com/haukot/copilot_auth_example

**Comment by zbirenbaum** [comment]
> So, if it'd help with auth, I managed to get it, you can check the example here https://github.com/haukot/copilot_auth_example

This is fantastic, awesome work! I'll play around with it and see what I can get up and running

**Comment by zbirenbaum** [comment]
> So, if it'd help with auth, I managed to get it, you can check the example here https://github.com/haukot/copilot_auth_example

Very early WIP but I figured I'd post here since I just got the requests to copilot working using actual prefix and suffix information from the LSP sync'd document info, and its getting close to a state where people might want to contribute.

Right now it does send completion info to the editor, but I haven't gotten it formatted the same as the existing copilot LSP (can be tested by printing the table of completions returned by 'getCompletionsCycling' requests), and it's missing some info, so it won't appear properly (or at all) in your completions. Very much a cool toy atm, but I'll keep hacking away at it.

https://github.com/zbirenbaum/copilot-rs/tree/master
